File: qyz_chapter_six/src/main/java/com/example/qyz_chapter_six/MainActivity.java
package com.example.qyz_chapter_six;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;

/**
 * 双缓冲技术是游戏开发中的一个重要的技术。
 * 当一个动画争先显示时，程序又在改变它，前面还没有显示完，程序又请求重新绘制，这样屏幕就会不停地闪烁。
 * 而双缓冲技术是把要处理的图片在内存中处理好之后，再将其显示在屏幕上。
 * 双缓冲主要是为了解决 反复局部刷屏带来的闪烁。
 * 把要画的东西先画到一个内存区域里，然后整体的一次性画出来。
 *
 */
public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }
}

File: falling_ball_evaluator/src/main/java/com/example/falling_ball_evaluator/MainActivity.kt
package com.example.falling_ball_evaluator

import android.animation.ValueAnimator
import android.os.Bundle
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

/**
 * 小球抛物动画
 * 自定义Evaluator转换器将进度值转化为最终的value animator值
 */
class MainActivity : AppCompatActivity() {
    private var mCurPoint: Point? = null
    private var mIvCircle: ImageView? = null
    private var mTvReset :TextView ? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initView()
        bindEvent()
        initData()
    }

    private fun initData() {
        startAnimator()
    }

    private fun startAnimator() {
        val animator = ValueAnimator.ofObject(FullingBallEvaluator(), Point(0, 0), Point(500, 500))
        animator.addUpdateListener { animation: ValueAnimator ->
            mCurPoint = animation.animatedValue as Point
            mIvCircle!!.layout(mCurPoint!!.x, mCurPoint!!.y, mCurPoint!!.x + mIvCircle!!.width, mCurPoint!!.y + mIvCircle!!.height)
        }
        animator.duration = 2000

        animator.start()
    }

    private fun bindEvent() {
        mTvReset!!.setOnClickListener {
            startAnimator()
        }
    }

    private fun initView() {
        mIvCircle = findViewById(R.id.iv_circle)
        mTvReset = findViewById(R.id.tv_reset)
    }
}